Add notification_rule module

nicoske opened the pull request:
This is my tentative of implementation for a notification_rule module.

The module manages Checkmk notification rules via the REST API with full CRUD support:
- Create new notification rules
- Read and compare existing rules for idempotency
- Update rules when configuration changes are detected
- Delete rules

Tested with mail, slack and msteams notification plugins. The module is plugin-agnostic and supports all notification plugins by passing the rule_config directly to the API.

Includes integration tests.

nicoske committed:
Add notification_rule module

nicoske committed:
Add notification_rule module

robin-checkmk committed:
Add GitHub-related files.

robin-checkmk committed:
Align basic integration test logic with other modules.

robin-checkmk committed:
Align basic integration test logic with other modules.

robin-checkmk committed:
Bugfix integration test trigger for GitHub Actions.

robin-checkmk reviewed (changes requested):
I personally think, `notification` is the better name for this module. I think we do not need the `_rule` part, as that might create more confusion than merit. Technically we are of course managing notification rules, but I think from a collection user perspective it feels more natural (and easier for that matter) to user `checkmk.general.notification`.
What do you think?

robin-checkmk committed:
Remove EOL Checkmk 2.2.0 from integration tests.

robin-checkmk committed:
Remove EOL Checkmk 2.2.0 from integration tests.

nicoske commented:
> @nicoske wow, this is a massive task to take on, thank you so much! 🙏
> 
> As you might already have noticed: We can only work on this in our spare-time, so it might take a while to get this merged. But I might be wrong, maybe we can give users even yet another holiday present this year. :)
> 
> That being said: I will take a first look and provide initial feedback this week. Once initial findings have been addressed, I will hand over for final code review after which we will be able to merge.

No worries about the timing. I'm also doing this in my spare time and I take the opportunity to fill my own needs with this module.

nicoske commented:
> I personally think, `notification` is the better name for this module. I think we do not need the `_rule` part, as that might create more confusion than merit. Technically we are of course managing notification rules, but I think from a collection user perspective it feels more natural (and easier for that matter) to user `checkmk.general.notification`. What do you think?

I understand your argument. I picked `notification_rule` to match the API endpoint and avoid potential confusion with the notifications themselves, but I'd be happy to adapt.

robin-checkmk commented:
> I understand your argument. I picked `notification_rule` to match the API endpoint and avoid potential confusion with the notifications themselves, but I'd be happy to adapt.

Then please go ahead and work on my comments (as far as you agree, if not, let's discuss in the respective threads).
If anything comes up, I am here to help.
The naming question will be looked at by at least a second set of eyes, once we go for code review.

nicoske committed:
Move rule_config to vars

nicoske committed:
Move rule_config to vars

nicoske committed:
Add documentation and msteams example

nicoske committed:
Add documentation and msteams example

nicoske commented on plugins/modules/notification_rule.py:
```diff
@@ -0,0 +1,528 @@
+        required: false
+        type: str
+
+    rule_config:
```
Added documentation and msteams example. I'll try to add more once I tested them.

robin-checkmk replied:
The `contact_selection` and `conditions` options should all have defaults, so it should not be necessary to always explicitly mention them as in the examples currently, right?

nicoske replied:
The module now auto-fills missing contact_selection and conditions options with state: disabled.

robin-checkmk commented:
@nicoske thank you for the continuous work on this!

Would you mind to first rename the module to `notification`? That would be the blocker for merging for me.
The documentation can still be finalized/improved, after the feature was initially merged. :)

nicoske committed:
Rename module to notification

nicoske committed:
Rename module to notification

robin-checkmk committed:
Align GitHub Actions with new module name.

nicoske commented:
@robin-checkmk While testing the module I noticed it creates "notification parameter" entries named "Auto-generated during rule creation via REST API". Each unique parameter combination creates a new entry, so you end up with orphaned entries over time when you modify rules. I looked over CheckMK API and it seems it's hardcoded, it doesn't support referencing existing parameters. However there's a separate API at `/domain-types/configuration_entity/` for managing notification parameters directly (create, rename, delete) though I'm not sure how we could handle this (as a separate module  ?)

robin-checkmk commented on plugins/modules/notification.py:
```diff
@@ -0,0 +1,756 @@
+"""
+
+EXAMPLES = r"""
+# Create an HTML email notification rule
```
Of course, this is not the place in the code where it happens, but in my tests, rule creation is not idempotent. Is this known to you, or might I be holding it wrong?

robin-checkmk committed:
Add basic demo playbooks.

robin-checkmk reviewed (commented):
I found some idempotency issues, maybe you can look into those or comment?

In good news: The issue with the notification parameters is resolved with Checkmk 2.4.0p14 (https://checkmk.com/werk/18290). This means, that we do not have to worry about that anymore.

robin-checkmk committed:
Add changelog.

nicoske commented on plugins/modules/notification.py:
```diff
@@ -0,0 +1,756 @@
+    automation_user: "myuser"
+    automation_secret: "mysecret"
+    rule_id: "{{ notification_result.content.id }}"
+    state: "absent"
```
Indeed without rule_id it creates a new rule each time. The rule module has the same problem and attempts to find rules by comparing config but its docs warn this is unreliable. Any change to the rule config would break the match and create a duplicate anyway. Given notification rules have even more complex nested structures I don't think implementing the same approach would give better results.

robin-checkmk replied:
Hm. I do see your point, while at the same time I want to resist and ask for idempotency. But as with the `rule` module, the topic is tricky and while people might dislike it, I'd rather have a stable solution, which needs more work on user-side, than an unstable solution which will cause issues time and again.

So for deletion, I think we are fine, but we still need to address the creation of rules.

robin-checkmk replied:
I have mixed feelings about that.
How would you handle e.g., a situation where two rules have the same description?
Do not get me wrong, it would be an alternative, especially, if you can build it in a way, where the user can provide either option per their preference. Or you look for the rule_id and fall back to description with maybe a warning printed.
I am just worried, that users might expect the description not to be an ID and run into issues there. But we can probably work with documentation there.

So TL;DR: I think using the description if no rule_id is provided sounds reasonable. If you do the coding, I will do the documentation.

nicoske committed:
Auto-fill disabled state for unspecified options

nicoske committed:
Auto-fill disabled state for unspecified options

robin-checkmk committed:
Move module documentation to dedicated file.

nicoske committed:
Add plugin_params defaults and fix NoneType error

- Add defaults for mail, slack, msteams plugin_params
- Fix NoneType error when contact_selection/conditions is null
- Add integration test for minimal config

nicoske committed:
Add plugin_params defaults and fix NoneType error

- Add defaults for mail, slack, msteams plugin_params
- Fix NoneType error when contact_selection/conditions is null
- Add integration test for minimal config

nicoske commented on plugins/modules/notification.py:
```diff
@@ -0,0 +1,825 @@
+            choices=["present", "absent"],
+            required=True,
+        ),
+    )
```
Fixed the 'NoneType' object is not iterable error. It occurred when contact_selection or conditions were explicitly set to null.

nicoske committed:
Fix black formatting

nicoske committed:
Fix black formatting

robin-checkmk commented:
@nicoske I just received word, that the required fields for the notification endpoint have fixed in the latest 2.4 and 2.3 releases (and for the upcoming 2.5 release). Do you want to take a look at those changes to potentially simplify the code?
That might also make it more feasible to not require users to explicitly define the `state`  key, but implicitly do so.

robin-checkmk committed:
Integrate changes from `devel` branch.

nicoske committed:
Simplify module for Checkmk >= 2.3.0p42 API changes.

Remove merge_with_defaults and all DEFAULT_* constants as the API
now fills in defaults server-side. Add description-based rule lookup
for idempotent create/update/delete without requiring rule_id.
Use recursive subset comparison for change detection with minimal configs.

nicoske committed:
Simplify module for Checkmk >= 2.3.0p42 API changes.

Remove merge_with_defaults and all DEFAULT_* constants as the API
now fills in defaults server-side. Add description-based rule lookup
for idempotent create/update/delete without requiring rule_id.
Use recursive subset comparison for change detection with minimal configs.

robin-checkmk committed:
Hotfix pinned 2.5 version.

robin-checkmk committed:
Bump Checkmk versions.

robin-checkmk reviewed (changes requested):
I had an LLM review the module and it found some things.
I will also push some fixes to other files myself later today.
But if you want to look at `notification.py` right away, go ahead.
After functional testing on my end, I will leave a final review and once you went through these last findings, I will stage this PR for merge and release.
We are certainly at a stage, where this creates enough value, even if bugs might be found at a later time.

robin-checkmk committed:
Align integration tests with `devel`.

nicoske committed:
Fix add HTTP code mappings, parameter validation and fix content decoding

robin-checkmk committed:
Update documentation.

robin-checkmk committed:
Remove Checkmk 2.3 from tests and extend changelog.

robin-checkmk committed:
Last minute updates to the module documentation.

robin-checkmk committed:
Fix Checkmk versions.

robin-checkmk committed:
Fix duplicate variable.

lgetwan reviewed (approved):
I really like the way you implemented the module!
There's only a small nit pick of the cosmetic kind.

nicoske committed:
Fix verify_parameters error message to be state-agnostic.

robin-checkmk committed:
Merge branch 'devel' into feature/add-notification-rule-module

robin-checkmk commented on plugins/modules/notification.py:
```diff
@@ -0,0 +1,482 @@
+- When I(rule_id) is not provided, the module will try to find an existing rule
+  by matching the I(description) field in I(rule_properties).
+  If multiple rules match the same description, the module will fail and ask for a unique C(rule_id).
+- Requires Checkmk >= 2.3.0p42 or >= 2.4.0p22 for minimal configuration support.
```
I just cleaned it up.
As Checkmk 2.5 is on the horizon, and we reworked the whole notification system in 2.4, I think it is fair to only support 2.4 and newer.

robin-checkmk committed:
Finalize changelog and documentation.

robin-checkmk committed:
Bump Checkmk versions.
